### PR TITLE
Implement backward compatible mechanism for deprecated package tool

### PR DESCRIPTION
**Pull request by chjinche**
# Description

Up til now, we have several tool breaking caused by package changing or tool naming, which may makes user previous flow fail to run when they upgrade tools. To provide smooth experience, we implement a tool transition mechanism for backward compatibility.

In this PR, we do the following things:
- add `deprecated_tools` field to tool contract, which collects all previous tool ids to current tool.
- handle backward compatibility of package tool ID changes in loading tool.
- support to collect new tool id if node source tool is a legacy tool.

# All Promptflow Contribution checklist:
- [X] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [X] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [X] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_core/tools_manager.py [hunk 1]
```diff
@@ -404,6 +409,8 @@ class ToolLoader:
     def __init__(self, working_dir: str, package_tool_keys: Optional[List[str]] = None) -> None:
         self._working_dir = working_dir
         self._package_tools = collect_package_tools(package_tool_keys) if package_tool_keys else {}
+        # Used to handle backward compatibility of tool ID changes.
+        self._tool_id_mapping = build_tool_id_mapping(self._package_tools)
```
Comment (thy09) [inline src/promptflow/promptflow/_core/tools_manager.py:413]: Maybe `_transitioning_tools`? which is more straightforward with "transition_from"

Comment (thy09) [reply src/promptflow/promptflow/_core/tools_manager.py:413]: Or `_deprecated_tools`

Comment (chjinche) [reply src/promptflow/promptflow/_core/tools_manager.py]: Changed.

**Review by DaweiCai — APPROVED** [review]

File: src/promptflow/promptflow/_utils/tool_utils.py [hunk 2]
```diff
@@ -302,6 +302,25 @@ def load_function_from_function_path(func_path: str):
         )
 
 
+# Handling backward compatibility and generating a mapping between the previous and new tool IDs.
+def _find_deprecated_tools(package_tools) -> Dict[str, str]:
+    _deprecated_tools = {}
+    for tool_id, tool in package_tools.items():
+        # "deprecated_tools" is a list of old tool IDs that are mapped to the current tool ID.
+        if tool and "deprecated_tools" in tool:
```
Comment (thy09) [inline src/promptflow/promptflow/_utils/tool_utils.py:310]: Put the key as a constant?

Comment (thy09) [reply src/promptflow/promptflow/_utils/tool_utils.py:310]: Or maybe we can deserialize the json here

**Review by thy09 — APPROVED** [review]

**Review by melionel — APPROVED** [review]

**Pull request merged**